### ISSUE TITLE
Initial state "catch-up" problem

### ISSUE THREAD

**Issue by ajohnstonTE** [matrix-org/matrix-js-sdk#3648]
TL;DR: If a user joins and leaves a room, then refreshes the window, the matrix client appears to need to "catch up" and initially reports that the user is still in the room before updating shortly after that they have left the room. Unknown how to tell when this initial sync process is truly complete or how to avoid the problem.

TL: If a user joins a room and leaves the room, then refreshes the window, the matrix client while starting up will emit multiple `ClientEvent.Sync` events. For one of those events, the room in question will have a timeline that _does not include_ the most recent leave event, such that the room appears to be one that the user is still a member of (both events/timeline-wise and on the "my membership" property of the room). Afterwards, another sync event will fire that has the recent leave event, and the user is correctly identified as not being a member of the room. The initial steps of joining and leaving a room can be repeated more than once such that it leads to more than just two `ClientEvent.Sync` events being fired, one for each join and each leave as it "catches up" with each event. This continues to happen for every page refresh for a few minutes, until eventually whatever is causing this seems to refresh its cache/etc and it only sends the full list of events that includes the most recent leave event.

It's entirely possible that I'm doing something wrong, but if so I don't know what. I'm only using `client.leave` (as well as `client.forget` but it happens even without that call) to leave the room, and `client.getRooms` to get the room list. `client.isInitialSyncComplete()` returns true for this entire duration (when logged during `Sync` events), so I don't know how I am supposed to know when the client has fully caught up on all events for a room.

Any help would be greatly appreciated.

**Comment by t3chguy** [comment]
So the issue is the sync accumulator is only persisted every 5 minutes, so if you refresh you will lose up to 5 minutes, so the 1st sync will be loading the sync accumulator, the next sync will be a catch-up sync. If you do not want this behaviour I suggest ensuring you fire `store.reallySave()` before the the app is closed to write the sync accumulator to storage.

**Comment by ajohnstonTE** [comment]
Thanks for the reply!

What is `store` referring to? Is that in the JS SDK?

**Comment by ajohnstonTE** [comment]
Ah, ok, I assume you mean the store from `client.store`. At the time we weren't using Indexeddb, we were just using the default memory store. Switching to IndexedDB + `client.store.save(true)` completely fixed the problem. Thanks!!!
